### PR TITLE
Added support for new offheap-resource addition in OffheapResourceProvider

### PR DESCRIPTION
**Pull request by saurabhagas**

### CODE REVIEW

File: offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourcesProvider.java [hunk 1]
```diff
@@ -158,13 +136,66 @@ static BigInteger convert(BigInteger value, MemoryUnit unit) {
...
+          updateCurrentUsage(newCapacity - oldCapacity);
+          for (EntityManagementRegistry registry : registries) {
+            Map<String, String> attrs = new HashMap<>();
+            attrs.put("oldCapacity", Long.toString(oldCapacity));
+            attrs.put("newCapacity", Long.toString(newCapacity));
+            registry.pushServerEntityNotification(res.getManagementBinding(), "OFFHEAP_RESOURCE_CAPACITY_CHANGED", attrs);
+          }
+        });
+
+    resources.put(identifier, offHeapResource);
```
Comment (chaktatha) [inline offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourcesProvider.java]: There may be subtle race here. We should use putIfAbsent. If putIfAbsent does not return null, return false.

File: offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourcesProvider.java [hunk 2]
```diff
@@ -158,13 +136,66 @@ static BigInteger convert(BigInteger value, MemoryUnit unit) {
...
+        "allocatedMemory",
+        new HashSet<>(Arrays.asList("OffHeapResource", "tier")),
+        properties,
+        StatisticType.GAUGE,
+        () -> offHeapResource.capacity() - offHeapResource.available()
+    );
+    return true;
+  }
+
+  private void updateCurrentUsage(long delta) {
```
Comment (chaktatha) [inline offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourcesProvider.java]: Considering we are only printing a warning message even when totalConfiguredOffheap > physicalMemory, this is acceptable.

File: offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourcesProvider.java (hunk 1 above)
Comment (saurabhagas) [reply offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourcesProvider.java]: Done

File: offheap-resource/src/test/java/org/terracotta/offheapresource/OffHeapResourcesProviderTest.java [hunk 3]
```diff
@@ -58,64 +67,73 @@ public void testObserverExposed() {
...
+  }
+
+  @Test
+  public void testConcurrentOffheapAddition() throws Exception {
+    long perThreadIncrement = 100L;
+    int updatesPerThread = 50;
+    int numThreads = 20;
+    ExecutorService executorService = Executors.newFixedThreadPool(numThreads);
+    OffHeapResourcesProvider offHeapResourcesProvider = new OffHeapResourcesProvider(configuration);
+    executorService.execute(() -> {
```
Comment (ramsai1729) [inline offheap-resource/src/test/java/org/terracotta/offheapresource/OffHeapResourcesProviderTest.java]: is this a concurrent test?

Comment (saurabhagas) [reply offheap-resource/src/test/java/org/terracotta/offheapresource/OffHeapResourcesProviderTest.java]: Oops! Raised PR#557 to fix it. Thanks!

**Pull request merged**